Kubernetes cluster hardening standard (previously "K8s cluster baseline security") (#581)
SovereignCloudStack/standards: Standards/scs-0217-v1-cluster-hardening.md

This option enables Kubernetes components to authenticate as a user without providing a password,
which is neither possible for Kubernetes components nor planned in future releases.
This method is recommended in order to implement ACL for different Kubernetes components and
not give the Kubernetes API full root access to the etcd instance; instead, a separate user can be created.

#### TLS communication

etcd should use TLS for peer- and cluster-communication, so that traffic between different peered etcd instances as well
as the communication with the Kubernetes cluster can be secured.
etcd provides options for all these scenarios, including `--peer-key-file=peer.key` and `--peer-cert-file=peer.cert`
for securing peer communication and the flags `--key-file=k8sclient.key` and `--cert-file=k8sclient.cert` for securing
client communication (and therefore cluster communication).
Additionally, HTTPS should be used as the URL schema.
It is also possibly to use a separate CA for the etcd in order to separate and better control access through client
certificates, since etcd by default trusts all the certificates issued by the root CA [2][nsa-cisa].

> More information about authentication via TLS are provided in the chapter [ACL restrictions](#acl-restrictions).

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

### Securing endpoints

Kubernetes provides a well-defined set of ports in its default configuration. These ports are
used for inter-component communication as well as external access. Due to the distribution of
and information about Kubernetes clusters, it is easy for a bad actor to identify a clusters
ports and try to attack them. In order to minimize the attack vector, internal ports (and therefore components)
should not be accessible from external networks, except if there are requirements to enable this behavior.

A good way to restrict access would be a combination of firewalls with port

> blocking possible network separation.

Author: Yes thank you. Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

How this is done is highly dependent on the specific setup of the provider.
An additional document could be provided in the future to give basic
guidelines for this task.

A list of the default ports used in Kubernetes as well as the components accessing them can be found below:

#### Control plane nodes

| Ports     | Protocol | Purpose                 | Used by              | Type               |
|-----------|----------|-------------------------|----------------------|--------------------|
| 6443      | TCP      | API server              | All                  | External, internal |
| 2379-2380 | TCP      | etcd server             | kube-apiserver, etcd | Internal           |
| 10250     | TCP      | Kubelet API             | Self, Control plane  | Internal           |

> | 10255     | TCP      | Read-only Kubelet API   | (Monitoring)         | External, Internal |

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

| 10257     | TCP      | kube-controller-manager | Self                 | Internal           |
| 10259     | TCP      | kube-scheduler          | Self                 | Internal           |

#### Worker nodes

| Ports       | Protocol | Purpose               | Used by             | Type               |
|-------------|----------|-----------------------|---------------------|--------------------|

> | 10250       | TCP      | Kubelet API           | Self, Control plane | Internal           |

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889) and put sentence in to explain `Self` here.

> | 10255       | TCP      | Read-only Kubelet API | (Monitoring)        | External, internal |

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

| 30000-32767 | TCP      | NodePort Services     |                     | External           |

### API security, authentication and authorization

In order to secure Kubernetes against bad actors, limiting and securing access to API requests
is recommended, since requests to those are able to control the entire Kubernetes cluster.
Access control is applied to both human users and Kubernetes service accounts, which goes through
several stages after a request reaches the API.

> 1. The Kubernetes API server listens on port 6443 on the first non-localhost network interface by default,

Author: Hm good point. I just took the documentation from [here](https://kubernetes.io/docs/concepts/security/controlling-access/) for granted, but the [kube-api server docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) tell a different story, since `--bind-address` mentions
```
--bind-address string     Default: 0.0.0.0
	

The IP address on which to listen for the --secure-port port. The associated interface(s) must be reachable by the rest of the cluster, and by CLI/web clients. If blank or an unspecified address (0.0.0.0 or ::), all interfaces and IP address families will be used.
```
which would put the `kube-api` per default on ALL interfaces.

I guess this is somewhat contradicting and could require an upstream issue.

The request must contain the username of the requester, the requested action and the affected object.
Kubernetes supports different authorization modules like ABAC, RBAC or Webhooks. Only one of these
modules need to approve the request in order for it to be authorized.
4. The last step are Admission control modules, which can modify or reject requests after accessing
the objects contents.

#### Authentication

Kubernetes provides different internal authentication mechanisms, that can be used depending
on the requirements of the cluster provider and user. Multiple authentication systems can
be enabled and the [Kubernetes documentation][kubernetes-auth] recommends at least using two methods,
including Service Account Tokens and another method. Methods directly provided by Kubernetes include
the following (a more complete or up-to-date list may be found in the [Kubernetes authentication docs][kubernetes-auth]):

- *Static Token Files*

  This method reads bearer tokens from requests and checks them against a CSV file provided to Kubernetes containing
  three columns named `token`, `username` and `uid`. These tokens last indefinitely and the list cant be changed without a

>   restart of the API server. This makes this option unsuitable for productive clusters.

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

- *Service Account Tokens*

  A service account is an authenticator that uses signed bearer tokens for request verification.
  Service accounts can be given to the API server with a file containing PEM-encoded X509 RSA or
  ECDSA private or public keys that verify the Service Account Tokens.
  Service Accounts are normally created automatically by the API server and associated with the
  pods through the `ServiceAccount` admission controller. Tokens are signed JSON Web Tokens,
  that can be used as a Bearer Token or mounted into the pods for API server access.
  Since Service Account Tokens are mainly used to allow workloads accessing the API server,

>   they're not really intended to authenticate users in productive clusters.</p>

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

As outlined, most internal authentication mechanisms of Kubernetes aren't really
usable in productive environments at the current time. Instead, external authentication
should be used in order to provide production-ready workflows.
The Kubernetes documentation lists a few examples for external authenticators, e.g.

- [OpenIDConnect][openidconnect]
- Bearer Tokens with [Webhook Token Authentication][webhook-token]
- Request Header Authentication with an [Authenticating Proxy][authenticating-proxy]

All of these examples are useful to set up for an organization or can be used with
an already in-place solution. More information can be found in their respective
part of the Kubernetes documentation.
Most of these are good solutions for productive setups, since they enable easy
user management, access revocation and things like short-lived access tokens.
What will be used by your organization depends on the present setup and the use case.

#### Authorization

Authorization is done after the authentication step in order to check the rights
of a user within the system. Kubernetes authorizes API requests with the API server,

> which evaluates requests against all policies in place and then allow or denies these requests.

Author: Updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

customer-owner-relationship-like structures in place.
Nonetheless, neither ABAC, RBAC nor Webhook authorization can be recommended over the
other, since this all depends on the use case and required structure of a deployment.
Using at least one of these modes is recommended.

It is also recommended to enable the Node authorizer in order to limit Kubelet
permissions to a minimum operational state.

#### Admission Controllers

Admission controllers intercept requests to the Kubernetes API after the
authentication and authorization step, which validate and/or mutate the request.
This step is limited to `create`, `modify` and `delete` objects as well as custom
verbs, but other requests are not blocked.
Kubernetes provides multiple admission controllers, some of which are enabled by default.

One recommended admission controller is the [`NodeRestriction` controller][node-restriction],
which limits the `Node` and `Pod` objects a Kubelet is allowed to modify to their own `Node` or
objects that are bound to them. It also disallows updating or removing taints and prevents changing
or adding labels with a `node-restriction.kubernetes.io/` prefix.

> Be aware that Kubelets will only be limited by this admission controller, if the use credentials
> in the `system:nodes` group with a `system:node:<nodeName>` username. Administrators must therefore

Reviewer: ```suggestion
Be aware that Kubelets will only be limited by this admission controller, if the user credentials
in the `system:nodes` group begin with a `system:node:<nodeName>` username. Administrators must therefore
```
not sure this is the intended meaning?

Author: Yes, thanks for the catch.
[6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

server and ensures, that pods handed over to it are running and healthy according
to the specification provided to it. The HTTPS endpoint of a Kubelet exposes APIs
with varying access to sensitive data and also enables various levels
of performant operations enabling manipulation of node data and containers.
There is also a read-only HTTP endpoint that was used for monitoring a Kubelet and
its information. This port was also used by applications like `kubeadm` to check
the health status of the Kubelet.
This port is still available, but it is planned to be [removed][ro-port-removal]
in a future version. At the moment, the port is disabled by default since [Kubernetes 1.10][ro-port-disabled]
and shortly later also in [`kubeadm`][ro-port-disabled-kubeadm].
Different sources recommend disabling this port [3][ro-port-s1] [4][ro-port-s2] due to possible
security risks, but since this standard recommends restricting accessibility of internal ports,
this port wouldn't be accessible from external networks.
It is nevertheless recommended to keep this port disabled, since Kubernetes also acknowledged
its risks and plans to remove it.

> By default, requests to the HTTPS endpoint that are not rejected by other authentication

Reviewer 2: We can also mention `By default, the API server does not verify the kubelet's serving certificate` here - see https://kubernetes.io/docs/concepts/architecture/control-plane-node-communication/#api-server-to-kubelet

Author: Thanks for the catch, updated in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

If this is the case, access to these instances MUST be configured, so that only the API server and
necessary cluster components requiring access can access etcd.
Communication with etcd MUST be secured with TLS for both peer- and cluster-communication.
It is RECOMMENDED to use a CA separate from the one used for the Kubernetes cluster for etcd in
order to better control and issue certificates for clients allowed to access etcd.
ACL MUST be enabled for etcd, which allows better control of the access rights to specific key sets
for specific users. Authentication MUST be done via the Common Name (CN) field of the TLS client
certificates (since normal username-password-authentication isn't implemented for Kubernetes).

Kubernetes' endpoints MUST be secured in order to provide a small attack surface for bad actors.
It MUST NOT be possible to access Kubernetes ports from outside the internal network hosting the
Kubernetes cluster except for the ports of the API server (default 6443) and the NodePort Services
(default 30000-32767). It is also possible to give access to the read-only Kubelet API port (default 10255),

> which is mostly used for monitoring, but we advise against this, mainly because the port is HTTP-only

Reviewer: why mention that this can be enabled if we advise against it's usage? this seems contradicting to me.

Author: Yeah you're right.
Removed the mention and updated the whole sentence to make it fit better in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

and can deliver sensitive information to the outside. It is RECOMMENDED to disable this port altogether,
if it isn't in use. Endpoints MUST be secured via HTTPS.

Securing Kubernetes via authentication and authorization is another important topic here.
Authentication is possible through multiple mechanisms, including Kubernetes-provided systems as well as external
authentication processes.
A cluster MUST implement at least two methods for authentication. One of these MUST be *Service Account Tokens*, in order

> to provide full functionality to Pods. A second authentication mechanisms can be chosen depending

Reviewer: I think the end of this sentence is missing? depending on what?

Author: Updated the sentence in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

Authorization also can be provided through multiple mechanisms.
A cluster MUST activate at least two authorization methods, one of which MUST be *Node authorization* and another one
consisting of either ABAC, RBAC or Webhook authorization depending on the required use case.
We RECOMMEND RBAC due to it fitting most use cases and being very well documented, but your setup might require another solution.

In order to harden Kubelet access control, a Kubelet SHOULD only be accessible internally via HTTPS. This is already the
case for the Kubelet API, but the read-only port is only available as HTTP. It is up to the provider to decide if this

> read-only port is required or if it can be deactivated (do we recommend this?).

Reviewer: please remove the text in the brackets.

Author: Removed in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889)

Kubelets MUST disable anonymous request authentication to disallow non-rejected requests to go through as anonymous requests.
OPTIONALLY, X509 client certificate authentication or API bearer token authentication can be enabled.
Request authorization for the Kubelet MUST be delegated to the API server via `Webhook` authorization as it is recommended
by the [Kubernetes documentation][kubelet-auth].
Additionally, the `NodeRestriction` admission controller MUST be activated in order to limit interactions between

> different Kubelets by disallowing modification of non-bound.

Reviewer: this sentence is also missing some words I believe? non-bound what?

Author: Updated this section, so this sentence now finishes (in [6805ea0](https://github.com/SovereignCloudStack/standards/pull/581/commits/6805ea0289d5595323c73ed5df8b29d9e90fd889))

[node-restriction]: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction
[kubelet-auth]: https://kubernetes.io/docs/reference/access-authn-authz/kubelet-authn-authz/#kubelet-authorization
[pod-security-standards]: https://kubernetes.io/docs/concepts/security/pod-security-standards/
[openidconnect]: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens
[webhook-token]: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication
[authenticating-proxy]: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy

[ro-port-removal]: https://github.com/kubernetes/kubernetes/issues/12968
[ro-port-disabled]: https://github.com/kubernetes/kubernetes/pull/59666
[ro-port-disabled-kubeadm]: https://github.com/kubernetes/kubeadm/issues/732
[ro-port-s1]: https://www.stigviewer.com/stig/kubernetes/2021-04-14/finding/V-242387
[ro-port-s2]: https://docs.datadoghq.com/security/default_rules/cis-kubernetes-1.5.1-4.2.4/
[nsa-cisa]: https://kubernetes.io/blog/2021/10/05/nsa-cisa-kubernetes-hardening-guidance/

> [etcd-auth]: https://etcd.io/docs/v3.3/op-guide/authentication/

Reviewer: we should probably include a linter in our pipeline which can check if all these links still work, else we will suffer from link rot in the future.

Reviewer 3: Link rot won't happen if we use permalinks (/sarcasm)

Author: Im gonna resolve this, the previously mentioned issue will be worked on by @martinmo

etcd can also be launched with `--client-cert-auth=true`, which enables authentication via
the Common Name (CN) field of a client TLS certificate without a password.
This option enables Kubernetes components to authenticate as a user without providing a password,
which is neither possible for Kubernetes components nor planned in future releases.
This method is recommended in order to implement ACL for different Kubernetes components and
not give the Kubernetes API full root access to the etcd instance; instead, a separate user can be created.

#### TLS communication

etcd should use TLS for peer- and cluster-communication, so that traffic between different peered etcd instances as well
as the communication with the Kubernetes cluster can be secured.
etcd provides options for all these scenarios, including `--peer-key-file=peer.key` and `--peer-cert-file=peer.cert`
for securing peer communication and the flags `--key-file=k8sclient.key` and `--cert-file=k8sclient.cert` for securing
client communication (and therefore cluster communication).
Additionally, HTTPS should be used as the URL schema.

> It is also possibly to use a separate CA for the etcd in order to separate and better control access through client

Reviewer 2: ```suggestion
It is also possible to use a separate CA for the etcd in order to separate and better control access through client
```

Author: Updated in a2adca5b18c82097388b59a88dc665c0d2b72ab4

certificates, since etcd by default trusts all the certificates issued by the root CA [2][nsa-cisa].
More information about authentication via TLS is provided in the chapter [ACL restrictions](#acl-restrictions).

### Securing endpoints

Kubernetes provides a well-defined set of ports in its default configuration. These ports are
used for inter-component communication as well as external access. Due to the distribution of

> and information about Kubernetes clusters, it is easy for a bad actor to identify a clusters

Reviewer 2: ```suggestion
information about Kubernetes clusters, it is easy for a bad actor to identify a clusters
```

Author: Updated in a2adca5b18c82097388b59a88dc665c0d2b72ab4

3. The authorization step authorizes a request, if a user is allowed to carry out a specific operation.
The request must contain the username of the requester, the requested action and the affected object.
Kubernetes supports different authorization modules like ABAC, RBAC or Webhooks. Only one of these
modules need to approve the request in order for it to be authorized.
4. The last step are Admission control modules, which can modify or reject requests after accessing
the objects contents.

#### Authentication

Kubernetes provides different internal authentication mechanisms, that can be used depending
on the requirements of the cluster provider and user. Multiple authentication systems can
be enabled and the [Kubernetes documentation][kubernetes-auth] recommends at least using two methods,
including Service Account Tokens and another method. Methods directly provided by Kubernetes include
the following (a more complete or up-to-date list may be found in the [Kubernetes authentication docs][kubernetes-auth]):

- *Static Token Files*

  This method reads bearer tokens from requests and checks them against a CSV file provided to Kubernetes containing

>   three columns named `token`, `username` and `uid`. These tokens last indefinitely and the list cant be changed without a

Reviewer 2: ```suggestion
  three columns named `token`, `username` and `uid`. These tokens last indefinitely and the list can't be changed without a
```

Author: Updated in a2adca5b18c82097388b59a88dc665c0d2b72ab4

distroless images, usage of `ImagePolicyWebhook`, network policy enforcement,
container sandboxing and prevention of kernel module loading.
Most of these were taken out of the document during writing due to either being the responsibility
of the clusters' user (and therefore not possible to implement for the provider), being more relevant
for high security clusters or changing the expected cluster environment too much, so that normally
expected operations could potentially not work in such a modified cluster.
These measurements will possibly be introduced in a future document about higher security clusters.

## Standard

This standard provides the baseline security requirements for a cluster in the SCS context. It tries
to increase the security in contrast to a non-configured setup by requiring and recommending specific
settings that provide different security improvements without being too intrusive.

Kubernetes clusters MUST be updated regularly in order to receive bugfixes and security patches.
For more information refer to the [SCS K8s Version Policy](scs-0210-v2-k8s-version-policy.md),

> which outlines the update policies of the SCS.

Reviewer 2: ```suggestion
which outlines the updated policies of the SCS.
```

Author: Updated in a2adca5b18c82097388b59a88dc665c0d2b72ab4

Endpoints MUST be secured via HTTPS.

Securing Kubernetes via authentication and authorization is another important topic here.
Authentication is possible through multiple mechanisms, including Kubernetes-provided systems as well as external
authentication processes.
A cluster MUST implement at least two methods for authentication. One of these MUST be *Service Account Tokens*, in order
to provide full functionality to Pods. A second authentication mechanisms can be chosen depending on the requirements
of the provider and/or customer.

Authorization also can be provided through multiple mechanisms.
A cluster MUST activate at least two authorization methods, one of which MUST be *Node authorization* and another one
consisting of either ABAC, RBAC or Webhook authorization depending on the required use case.
We RECOMMEND RBAC due to it fitting most use cases and being very well documented, but your setup might require another solution.

In order to harden Kubelet access control, a Kubelet SHOULD only be accessible internally via HTTPS. This is already the
case for the Kubelet API, but the read-only port is only available as HTTP. It is up to the provider to decide if this

> read-only port is required or if it can be deactivated.

Reviewer 4: > but the read-only port is only available as HTTP. It is up to the provider to decide if this read-only port is required or if it can be deactivated.

Above, it is stated:

> The read-only Kubelet API port (default 10255), which is mostly used for monitoring,
SHOULD be disabled altogether

These two statements are IMO not directly in conflict, but removing one of them (preferably removing the one without the SHOULD) might improve things.

It's just a general observation, not tied to the HTTP read-only port. If you give a standard statement with a keyword from RFC 2119 regarding an aspect of the implementation (here: handling of the port), there should only be one single reference to this issue in the document.

Author: I made some small adjustments here, so that the second section kind of references the first one.

At last, *Pod security standards* in the form of policies MUST be activated for the cluster. The SCS REQUIRES at least
the *Baseline* policy with the *Restricted* policy CAN also be used.
The *Pod security* admission controller MUST also be activated in order to enforce these policies on a namespace level.
We RECOMMEND the `enforce` level to be used for this admission controller setup.

## Conformance Tests

Conformance Tests will be written within another issue.

## Related Documents

[OWASP Kubernetes Security Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Kubernetes_Security_Cheat_Sheet.html)
[Kubernetes security concepts](https://kubernetes.io/docs/concepts/security/)
[Securing a cluster](https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/)
[Controlling access](https://kubernetes.io/docs/concepts/security/controlling-access/)
[Pod security standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/)
[NSA CISA Kubernetes hardening](https://kubernetes.io/blog/2021/10/05/nsa-cisa-kubernetes-hardening-guidance/)
[Configure etcd](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/)

> [Google Kubernetes cluster trust](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-trust)

Author: Adjusted this part.
